### PR TITLE
Fix JNI build on FreeBSD

### LINKED ISSUE #1286: freebsd compile error
Hi.
Freebsd build fail with `stoull` is not a member of `std` error

```
]>uname -a
FreeBSD devel.ximad.com 10.2-RELEASE-p9 FreeBSD 10.2-RELEASE-p9 #0: Thu Jan 14 01:32:46 UTC 2016     root@amd64-builder.daemonology.net:/usr/obj/usr/src/sys/GENERIC  amd64
```

```
]>gcc -v
Using built-in specs.
COLLECT_GCC=gcc
COLLECT_LTO_WRAPPER=/usr/local/libexec/gcc48/gcc/x86_64-portbld-freebsd10.2/4.8.5/lto-wrapper
Target: x86_64-portbld-freebsd10.2
Configured with: ./../gcc-4.8.5/configure --disable-multilib --disable-bootstrap --disable-nls --enable-gnu-indirect-function --libdir=/usr/local/lib/gcc48 --libexecdir=/usr/local/libexec/gcc48 --program-suffix=48 --with-as=/usr/local/bin/as --with-gmp=/usr/local --with-gxx-include-dir=/usr/local/lib/gcc48/include/c++/ --with-ld=/usr/local/bin/ld --with-pkgversion='FreeBSD Ports Collection' --with-system-zlib --with-ecj-jar=/usr/local/share/java/ecj-4.5.jar --enable-languages=c,c++,objc,fortran,java --prefix=/usr/local --localstatedir=/var --mandir=/usr/local/man --infodir=/usr/local/info/gcc48 --build=x86_64-portbld-freebsd10.2
Thread model: posix
gcc version 4.8.5 (FreeBSD Ports Collection) 
```

```
]>git clone https://github.com/facebook/rocksdb
]>cd rocksdb
]>gmake
Makefile:101: Warning: Compiling in debug mode. Don't use the resulting binary in production
  GEN      util/build_version.cc
...
  CC       util/lru_cache.o
  CC       util/threadpool.o
  CC       util/transaction_test_util.o
util/transaction_test_util.cc: In member function 'bool rocksdb::RandomTransactionInserter::DoInsert(rocksdb::DB*, rocksdb::Transaction*, bool)':
util/transaction_test_util.cc:97:19: error: 'stoull' is not a member of 'std'
       int_value = std::stoull(value);
                   ^
util/transaction_test_util.cc: In static member function 'static rocksdb::Status rocksdb::RandomTransactionInserter::Verify(rocksdb::DB*, uint16_t)':
util/transaction_test_util.cc:208:28: error: 'stoull' is not a member of 'std'
       uint64_t int_value = std::stoull(value.ToString());
                            ^
gmake: *** [Makefile:1456: util/transaction_test_util.o] Error 1
```

### PR DESCRIPTION
**Pull request by avsej**
Also fixes #1286

### CODE REVIEW

File: build_tools/build_detect_platform [hunk 1]
```diff
@@ -134,7 +134,7 @@ case "$TARGET_OS" in
         ;;
     FreeBSD)
         PLATFORM=OS_FREEBSD
-        COMMON_FLAGS="$COMMON_FLAGS -fno-builtin-memcmp -D_REENTRANT -DOS_FREEBSD"
+        COMMON_FLAGS="$COMMON_FLAGS -fno-builtin-memcmp -D_REENTRANT -DOS_FREEBSD -D_GLIBCXX_USE_C99 -fpermissive"
```
Comment (adamretter) [inline build_tools/build_detect_platform]: Just a query as to whether `-D_GLIBCXX_USE_C99` is correct?

Comment (avsej) [inline build_tools/build_detect_platform]: no, without `-fpermissive` these warnings turned into errors

```
java/rocksjni/iterator.cc: In function '__jbyteArray* Java_org_rocksdb_RocksIterator_key0(JNIEnv*, jobject, jlong)':
java/rocksjni/iterator.cc:125:75: warning: invalid conversion from 'const jbyte* {aka const signed char*}' to 'jbyte* {aka signed char*}' [-fpermissive]
                           reinterpret_cast<const jbyte*>(key_slice.data()));
                                                                           ^
java/rocksjni/iterator.cc: In function '__jbyteArray* Java_org_rocksdb_RocksIterator_value0(JNIEnv*, jobject, jlong)':
java/rocksjni/iterator.cc:142:77: warning: invalid conversion from 'const jbyte* {aka const signed char*}' to 'jbyte* {aka signed char*}' [-fpermissive]
                           reinterpret_cast<const jbyte*>(value_slice.data()));
                                                                             ^
java/rocksjni/rocksjni.cc: In function '__jobjectArray* Java_org_rocksdb_RocksDB_listColumnFamilies(JNIEnv*, jclass, jlong, jstring)':
java/rocksjni/rocksjni.cc:197:72: warning: invalid conversion from 'const jbyte* {aka const signed char*}' to 'jbyte* {aka signed char*}' [-fpermissive]
           reinterpret_cast<const jbyte*>(column_family_names[i].data()));
                                                                        ^
java/rocksjni/rocksjni.cc: In function '__jbyteArray* rocksdb_get_helper(JNIEnv*, rocksdb::DB*, const rocksdb::ReadOptions&, rocksdb::ColumnFamilyHandle*, jbyteArray, jint, jint)':
java/rocksjni/rocksjni.cc:505:74: warning: invalid conversion from 'const jbyte* {aka const signed char*}' to 'jbyte* {aka signed char*}' [-fpermissive]
                             reinterpret_cast<const jbyte*>(value.c_str()));
                                                                          ^
java/rocksjni/rocksjni.cc: In function 'jint rocksdb_get_helper(JNIEnv*, rocksdb::DB*, const rocksdb::ReadOptions&, rocksdb::ColumnFamilyHandle*, jbyteArray, jint, jint, jbyteArray, jint, jint)':
java/rocksjni/rocksjni.cc:630:73: warning: invalid conversion from 'const jbyte* {aka const signed char*}' to 'jbyte* {aka signed char*}' [-fpermissive]
                           reinterpret_cast<const jbyte*>(cvalue.c_str()));
                                                                         ^
java/rocksjni/rocksjni.cc: In function '__jobjectArray* multi_get_helper(JNIEnv*, jobject, rocksdb::DB*, const rocksdb::ReadOptions&, jobjectArray, jintArray, jintArray, jlongArray)':
java/rocksjni/rocksjni.cc:708:60: warning: invalid conversion from 'const jbyte* {aka const signed char*}' to 'jbyte* {aka signed char*}' [-fpermissive]
           reinterpret_cast<const jbyte*>(values[i].c_str()));
                                                            ^
java/rocksjni/rocksjni.cc: In function '__jlongArray* Java_org_rocksdb_RocksDB_iterators(JNIEnv*, jobject, jlong, jlongArray, jlong)':
java/rocksjni/rocksjni.cc:1267:76: warning: invalid conversion from 'const jlong* {aka const long int*}' to 'jlong* {aka long int*}' [-fpermissive]
                               reinterpret_cast<const jlong*>(&iterators[i]));
                                                                            ^
java/rocksjni/slice.cc: In function '__jbyteArray* Java_org_rocksdb_Slice_data0(JNIEnv*, jobject, jlong)':
java/rocksjni/slice.cc:168:50: warning: invalid conversion from 'const jbyte* {aka const signed char*}' to 'jbyte* {aka signed char*}' [-fpermissive]
     reinterpret_cast<const jbyte*>(slice->data()));
                                                  ^
java/rocksjni/writebatchhandlerjnicallback.cc: In member function '__jbyteArray* rocksdb::WriteBatchHandlerJniCallback::sliceToJArray(const rocksdb::Slice&)':
java/rocksjni/writebatchhandlerjnicallback.cc:97:47: warning: invalid conversion from 'const jbyte* {aka const signed char*}' to 'jbyte* {aka signed char*}' [-fpermissive]
       reinterpret_cast<const jbyte*>(s.data()));
                                               ^
java/rocksjni/write_batch_test.cc: In function '__jbyteArray* Java_org_rocksdb_WriteBatchTest_getContents(JNIEnv*, jclass, jlong)':
java/rocksjni/write_batch_test.cc:105:72: warning: invalid conversion from 'const jbyte* {aka const signed char*}' to 'jbyte* {aka signed char*}' [-fpermissive]
                           reinterpret_cast<const jbyte*>(state.c_str()));
                                                                        ^
```

Comment (avsej) [inline build_tools/build_detect_platform]: oh, sorry, you are asking about `_GLIBCXX_USE_C99`, it is required, and rocksdb in ports also uses it

https://github.com/freebsd/freebsd-ports/commit/6adb4a5b5e4770e30d9d10fa14e152bf4c8607be

Comment (adamretter) [inline build_tools/build_detect_platform]: @avsej Okay cool. Let me bring up a FreeBSD VM to test this on, and all being well I will merge.

Comment (avsej) [inline build_tools/build_detect_platform]: I've signed individual CLA

File: build_tools/build_detect_platform [hunk 2]
```diff
@@ -134,7 +134,7 @@ case "$TARGET_OS" in
         ;;
     FreeBSD)
         PLATFORM=OS_FREEBSD
-        COMMON_FLAGS="$COMMON_FLAGS -fno-builtin-memcmp -D_REENTRANT -DOS_FREEBSD"
+        COMMON_FLAGS="$COMMON_FLAGS -fno-builtin-memcmp -D_REENTRANT -DOS_FREEBSD -D_GLIBCXX_USE_C99"
```
Comment (avsej) [inline build_tools/build_detect_platform]: @adamretter I removed `-fpermissive` and it does not actually affect the build (those warnings still treated as errors).
I've built it with `gmake rocksdbjava DISABLE_WARNING_AS_ERROR=1`

**Comment by adamretter** [comment]
@avsej So I am trying to reproduce this, I am not that familiar with building on FreeBSD, I used to use FreeBSD as my operating system but that was a few years ago now.

The process I am following on FreeBSD 11 is:

```
cd /usr/ports/devel/gflags
sudo make install

cd /usr/ports/archivers/bzip2
sudo make install

cd /usr/ports/archivers/snappy
sudo make install

(Optional for RocksJava)
cd /usr/ports/java/openjdk7
sudo make install
```

```
git clone https://github.com/facebook/rocksdb.git
cd rocksdb
CXX=clang++ CPPFLAGS="-I/usr/local/include -I/usr/local/openjdk7/include -I/usr/local/openjdk7/freebsd" CFLAGS="-I/usr/local/include -I/usr/local/openjdk7/include -I/usr/local/openjdk7/include/freebsd" LDFLAGS="-L/usr/local/lib" gmake -j4 rocksdbjava
```

Unfortunately this gives the error:
```
gmake[1]: Leaving directory '/usr/home/aretter/rocksdb/java'
/usr/bin/ld: cannot find -lsnappy
clang++: error: linker command failed with exit code 1 (use -v to see invocation)
gmake: *** [Makefile:1407: rocksdbjava] Error 1
```

I am not clear why it can't find libsnappy as I have added `/usr/local/lib` to the `LDFLAGS` above  can see it is present in `/usr/local/lib`:
```
$ ls /usr/local/lib/libsnappy*
/usr/local/lib/libsnappy.a		/usr/local/lib/libsnappy.so		/usr/local/lib/libsnappy.so.1		/usr/local/lib/libsnappy.so.1.3.0
```

What am I missing?

**Comment by adamretter** [comment]
@avsej Do you have any updates on my question please? If I don't hear back from you in about a week, I think we should close this ticket and mark as "abandoned".

**Pull request closed without merging**